### PR TITLE
Add a flag to switch caching on and off and factorize duplications

### PR DESCRIPTION
**Pull request by HerrHorizontal**
This PR adds an option to turn on/off the caching of input-files and also factorizes the workload creation and the duplication, in order to get dependent input-files.

This enables the execution of exactly the same jobs, without changing the initial hitrate of input-files on the caches.

### CODE REVIEW

**Comment by HerrHorizontal** [comment]
Might need some help here. The duplication doesn't quite work. I get a `SegFault` in the [`FileWriteAction`](https://github.com/HEPCompSim/DistCacheSim/blob/feature/switch-caching/src/SimpleExecutionController.cpp#L148-L151), which is probably related to the point, that I try to duplicate `shared_ptr`. But I don't get what exactly is going wrong.

**Comment by henricasanova** [comment]
Can you post a command-line for your simulation that makes this bug happen so that I can reproduce? Which command-line option triggers the feature and the bug?

**Comment by HerrHorizontal** [comment]
Sure, that's what I run:
```bash
sgbatch-sim --platform data/platform-files/ \
            --njobs 1 \
            --ninfiles 20 --insize $(bc -l <<< "8554379000 / ${NINFILES}") \
            --sigma-insize 10000000 \
            --hitrate 0 \
            --flops 1480000 \
            --sigma-flops 10000 \
            --mem 2400 \
            --sigma-mem 0 \
            --outsize 0 \
            --sigma-outsize 0 \
            --duplications 2 \
            --output-file test.csv \
            --no-caching
```
But I think it should be sufficient to reproduce the bug, when executing any command with `--duplications` set to any other value bigger than 1.

**Comment by ArturAkh** [comment]
@HerrHorizontal and @henricasanova,

I think the problem is solved:

https://github.com/HEPCompSim/DistCacheSim/pull/39/commits/7536546dde6e98cde785cd61c7429f6f1cb60f41

The problem was, that the worload specification was not returned by reference:

```diff
arturgo@arturgo-ThinkPad-P14s-Gen-2i:~/DistCacheSim$ git diff src/SimpleExecutionController.h
diff --git a/src/SimpleExecutionController.h b/src/SimpleExecutionController.h
index e151081..28d3fbf 100644
--- a/src/SimpleExecutionController.h
+++ b/src/SimpleExecutionController.h
@@ -33,7 +33,7 @@ public:
               //const double& hitrate,
               const std::string& outputdump_name);
 
-    std::map<std::string, JobSpecification> get_workload_spec() {
+    std::map<std::string, JobSpecification>& get_workload_spec() {
         return this->workload_spec;
     }
```

This lead to the problem, that the pointer to the location of the output file was always a 0 pointer.

Now this is fixed.

**Comment by HerrHorizontal** [comment]
I tested the fix and it works. Everything is fine now. I think this is ready for review.

File: src/SimpleExecutionController.cpp [hunk 1]
```diff
@@ -150,23 +150,23 @@ int SimpleExecutionController::main() {
...
+        // //         // TODO: one that's picked by the "user"?
+        // //         // TODO: Write the file at once
+        // //     },
+        // //     [](std::shared_ptr<wrench::ActionExecutor> action_executor) {
+        // //         WRENCH_INFO("Output file was successfully written!")
+        // //         // Do nothing
+        // //     }
+        // // );
+
+        // // Add necessary dependencies
```
Comment (ArturAkh) [inline src/SimpleExecutionController.cpp:169]: This can be removed completely

Comment (ArturAkh) [reply src/SimpleExecutionController.cpp:169]: Or at least the additional `//`

**Review by ArturAkh — APPROVED** [review]
Some comment markers can be removed, but we can do this in the process of merging into master. So fine for the time-being

**Pull request merged**